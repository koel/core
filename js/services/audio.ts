interface AudioService {
  context: AudioContext | null
  source: MediaElementAudioSourceNode | null
  element: HTMLMediaElement | null

  init (element: HTMLMediaElement): void
  getContext(): AudioContext
  getSource(): MediaElementAudioSourceNode
  getElement(): HTMLMediaElement
}

export const audio: AudioService = {
  context: null,
  source: null,
  element: null,

  init (element: HTMLMediaElement): void {
    const AudioContext = window.AudioContext ||
      window.webkitAudioContext ||
      window.mozAudioContext ||
      window.oAudioContext ||
      window.msAudioContext

    this.context = new AudioContext()
    this.source = this.context.createMediaElementSource(element)
    this.element = element
  },

  getContext () {
    return this.context!
  },

  getSource () {
    return this.source!
  },

  getElement () {
    return this.element!
  }
}
