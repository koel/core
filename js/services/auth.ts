import { ls } from '.'

export const auth = {
  storageKey: 'api-token',

  getToken (): string | null {
    return ls.get<string | null>(this.storageKey)
  },

  hasToken (): boolean {
    return Boolean(this.getToken())
  },

  setToken (token: string): void {
    ls.set(this.storageKey, token)
  },

  destroy (): void {
    ls.remove(this.storageKey)
  }
}
