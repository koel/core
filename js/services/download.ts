import { playlistStore, favoriteStore } from '@/stores'
import { ls } from '.'
import { alerts } from '@/utils'

interface Download {
  fromSongs(songs: Song | Song[]): void
  fromAlbum(album: Album): void
  fromArtist(artist: Artist): void
  fromPlaylist(playlist: Playlist): void
  fromFavorites(): void
  trigger(uri: string): void
}

let events: any

if (KOEL_ENV === 'app') {
  events = require('&/events').default
}

export const download: Download = {
  fromSongs (songs: Song | Song[]): void {
    const query = (<Song[]>[]).concat(songs).reduce((q, song) => `songs[]=${song.id}&${q}`, '')
    this.trigger(`songs?${query}`)
  },

  fromAlbum (album: Album): void {
    this.trigger(`album/${album.id}`)
  },

  fromArtist (artist: Artist): void {
    this.trigger(`artist/${artist.id}`)
  },

  fromPlaylist (playlist: Playlist): void {
    if (playlistStore.getSongs(playlist).length) {
      this.trigger(`playlist/${playlist.id}`)
    }
  },

  fromFavorites (): void {
    if (favoriteStore.all.length) {
      this.trigger('favorites')
    }
  },

  /**
   * Build a download link using a segment and trigger it.
   *
   * @param  {string} uri The uri segment, corresponding to the song(s),
   *                      artist, playlist, or album.
   */
  trigger: (uri: string) => {
    const sep = uri.includes('?') ? '&' : '?'
    const url = `${window.BASE_URL}api/download/${uri}${sep}jwt-token=${ls.get('jwt-token')}`

    if (KOEL_ENV === 'app') {
      require('electron').ipcRenderer.send(events.DOWNLOAD, url)
      alerts.success('Download started!')
    } else {
      const iframe = document.createElement('iframe')
      iframe.style.display = 'none'
      iframe.setAttribute('src', url)
      document.body.appendChild(iframe)
    }
  }
}
