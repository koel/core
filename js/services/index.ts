export * from './http'
export * from './info'
export * from './download'
export * from './ls'
export * from './playback'
export * from './youtube'
export * from './socket'
export * from './audio'
export * from './upload'
export * from './auth'
