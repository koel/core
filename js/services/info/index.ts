export { albumInfo } from './album'
export { artistInfo } from './artist'
export { songInfo } from './song'
