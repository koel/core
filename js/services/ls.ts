import { get as baseGet, set as baseSet, remove as baseRemove } from 'local-storage'

export const ls = {
  get: <T>(key: string, defaultValue: T | null = null): T | null => {
    const value = baseGet<T>(key)

    return value === null ? defaultValue : value
  },

  set: <T>(key: string, value: T): boolean => {
    return baseSet<T>(key, value)
  },

  remove: (key: string): void => {
    baseRemove(key)
  }
}
