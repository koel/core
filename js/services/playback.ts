import { shuffle, orderBy, throttle } from 'lodash'
import plyr from 'plyr'
import Vue from 'vue'
import isMobile from 'ismobilejs'

import { event, isMediaSessionSupported, isAudioContextSupported } from '@/utils'
import {
  queueStore,
  sharedStore,
  userStore,
  songStore,
  recentlyPlayedStore,
  preferenceStore as preferences
} from '@/stores'
import { socket, audio as audioService } from '.'
import { app } from '@/config'
import router from '@/router'

interface Playback {
  previous: Song | undefined
  next: Song | undefined
  isTranscoding: boolean
  player: Plyr | null
  volumeInput: HTMLInputElement | null
  repeatModes: string[]
  initialized: boolean
  mainWin: any

  init(): void
  listenToMediaEvents(media: HTMLMediaElement): void
  setVolume(value: number, persist?: boolean): void
  setMediaSessionActionHandlers(): void
  listenToSocketEvents(): void
  toggle(): void
  playNext(): void
  playPrev(): void
  play(song: Song | undefined): void
  restart(): void
  resume(): void
  pause(): void
  stop(): void
  mute(): void
  unmute(): void
  playFirstInQueue(): void
  queueAndPlay(songs?: Song[], shuffle?: boolean): void
  registerPlay(song: Song): void
  preload(song: Song): void
  showNotification(song: Song): void
  changeRepeatMode(): void
  playAllByArtist(artist: Artist, shuffle?: boolean): void
  playAllInAlbum(album: Album, shuffle?: boolean): void
}

/**
 * The number of seconds before the current song ends to start preload the next one.
 */
const PRELOAD_BUFFER = 30
const DEFAULT_VOLUME_VALUE = 7
const VOLUME_INPUT_SELECTOR = '#volumeRange'
const REPEAT_MODES = ['NO_REPEAT', 'REPEAT_ALL', 'REPEAT_ONE']

export const playback: Playback = {
  player: null,
  volumeInput: null,
  repeatModes: REPEAT_MODES,
  initialized: false,
  mainWin: null,

  init () {
    if (KOEL_ENV === 'app') {
      this.mainWin = require('electron').remote.getCurrentWindow()
    }

    // We don't need to init this service twice, or the media events will be duplicated.
    if (this.initialized) {
      return
    }

    this.player = plyr.setup(<HTMLMediaElement>document.querySelector('.plyr'), {
      controls: []
    })[0]

    this.volumeInput = document.querySelector(VOLUME_INPUT_SELECTOR)
    this.listenToMediaEvents(this.player.media)

    if (isAudioContextSupported) {
      try {
        this.setVolume(preferences.volume)
      } catch (e) {}

      audioService.init(this.player.media)
      event.emit(event.$names.INIT_EQUALIZER)
    }

    if (isMediaSessionSupported) {
      this.setMediaSessionActionHandlers()
    }

    // As of current, only the web-based version of Koel supports the remote controller
    if (KOEL_ENV !== 'app') {
      this.listenToSocketEvents()
    }

    this.initialized = true
  },

  listenToSocketEvents (): void {
    socket.listen(event.$names.SOCKET_TOGGLE_PLAYBACK, () => this.toggle())
      .listen(event.$names.SOCKET_PLAY_NEXT, () => this.playNext())
      .listen(event.$names.SOCKET_PLAY_PREV, () => this.playPrev())
      .listen(event.$names.SOCKET_GET_STATUS, () => {
        const data = queueStore.current ? songStore.generateDataToBroadcast(queueStore.current) : {
          volume: this.volumeInput!.value
        }
        socket.broadcast(event.$names.SOCKET_STATUS, data)
      })
      .listen(event.$names.SOCKET_GET_CURRENT_SONG, () => {
        socket.broadcast(
          event.$names.SOCKET_SONG,
          queueStore.current
            ? songStore.generateDataToBroadcast(queueStore.current)
            : { song: null }
        )
      })
      .listen(event.$names.SOCKET_SET_VOLUME, ({ volume } : { volume: number }) => this.setVolume(volume))
  },

  setMediaSessionActionHandlers (): void {
    // workaround for "TS2339: Property 'mediaSession' does not exist on type 'Navigator'" error
    ;(<any>navigator).mediaSession.setActionHandler('play', () => this.resume())
    ;(<any>navigator).mediaSession.setActionHandler('pause', () => this.pause())
    ;(<any>navigator).mediaSession.setActionHandler('previoustrack', () => this.playPrev())
    ;(<any>navigator).mediaSession.setActionHandler('nexttrack', () => this.playNext())
  },

  listenToMediaEvents (mediaElement: HTMLMediaElement): void {
    mediaElement.addEventListener('error', () => this.playNext(), true)

    mediaElement.addEventListener('ended', () => {
      if (sharedStore.state.useLastfm && userStore.current.preferences.lastfm_session_key) {
        songStore.scrobble(queueStore.current!)
      }

      preferences.repeatMode === 'REPEAT_ONE' ? this.restart() : this.playNext()
    })

    mediaElement.addEventListener('timeupdate', throttle((): void => {
      const currentSong = queueStore.current!

      if (!currentSong.playCountRegistered && !this.isTranscoding) {
        // if we've passed 25% of the song, it's safe to say the song has been "played".
        // Refer to https://github.com/koel/koel/issues/1087
        if (!mediaElement.duration || mediaElement.currentTime * 4 >= mediaElement.duration) {
          this.registerPlay(currentSong)
        }
      }

      const nextSong = queueStore.next

      if (!nextSong || nextSong.preloaded || this.isTranscoding) {
        return
      }

      if (mediaElement.duration && mediaElement.currentTime + PRELOAD_BUFFER > mediaElement.duration) {
        this.preload(nextSong)
      }
    }, 3000))
  },

  get isTranscoding (): boolean {
    return isMobile.any && preferences.transcodeOnMobile
  },

  registerPlay (song: Song): void {
    recentlyPlayedStore.add(song)
    songStore.registerPlay(song)
    recentlyPlayedStore.fetchAll()
    song.playCountRegistered = true
  },

  preload (song: Song): void {
    const audioElement = document.createElement('audio')
    audioElement.setAttribute('src', songStore.getSourceUrl(song))
    audioElement.setAttribute('preload', 'auto')
    audioElement.load()
    song.preloaded = true
  },

  /**
   * Play a song. Because
   *
   * So many adventures couldn't happen today,
   * So many songs we forgot to play
   * So many dreams swinging out of the blue
   * We'll let them come true
   */
  play (song: Song | undefined): void {
    if (!song) {
      return
    }

    document.title = `${song.title} ♫ ${app.name}`
    document.querySelector('.plyr audio')!.setAttribute('title', `${song.artist.name} - ${song.title}`)

    if (queueStore.current) {
      queueStore.current.playbackState = 'Stopped'
    }

    song.playbackState = 'Playing'
    queueStore.current = song

    // Manually set the `src` attribute of the audio to prevent plyr from resetting
    // the audio media object and cause our equalizer to malfunction.
    this.player!.media.src = songStore.getSourceUrl(song)

    // We'll just "restart" playing the song, which will handle notification, scrobbling etc.
    // Fixes #898
    if (isAudioContextSupported) {
      audioService.getContext().resume().then(() => this.restart())
    } else {
      this.restart()
    }
  },

  showNotification (song: Song): void {
    if (!window.Notification || !preferences.notify) {
      return
    }

    try {
      const notif = new window.Notification(`♫ ${song.title}`, {
        icon: song.album.cover,
        body: `${song.album.name} – ${song.artist.name}`
      })

      notif.onclick = () => KOEL_ENV === 'app' ? this.mainWin.focus() : window.focus()

      window.setTimeout(() => notif.close(), 5000)
    } catch (e) {
      // Notification fails.
      // @link https://developer.mozilla.org/en-US/docs/Web/API/ServiceWorkerRegistration/showNotification
      console.error(e)
    }

    if (isMediaSessionSupported) {
      ;(<any>navigator).mediaSession.metadata = new window.MediaMetadata({
        title: song.title,
        artist: song.artist.name,
        album: song.album.name,
        artwork: [
          { src: song.album.cover, sizes: '256x256', type: 'image/png' }
        ]
      })
    }
  },

  restart () {
    const song = queueStore.current!

    this.showNotification(song)

    // Record the UNIX timestamp the song starts playing, for scrobbling purpose
    song.playStartTime = Math.floor(Date.now() / 1000)

    song.playCountRegistered = false

    event.emit(event.$names.SONG_PLAYED, song)

    socket.broadcast(event.$names.SOCKET_SONG, songStore.generateDataToBroadcast(song))

    this.player!.restart()
    this.player!.play()
  },

  /**
   * The next song in the queue.
   * If we're in REPEAT_ALL mode and there's no next song, just get the first song.
   */
  get next (): Song | undefined {
    if (queueStore.next) {
      return queueStore.next
    }

    if (preferences.repeatMode === 'REPEAT_ALL') {
      return queueStore.first
    }
  },

  /**
   * The previous song in the queue.
   * If we're in REPEAT_ALL mode and there's no prev song, get the last song.
   */
  get previous (): Song | undefined {
    if (queueStore.previous) {
      return queueStore.previous
    }

    if (preferences.repeatMode === 'REPEAT_ALL') {
      return queueStore.last
    }
  },

  /**
   * Circle through the repeat mode.
   * The selected mode will be stored into local storage as well.
   */
  changeRepeatMode (): void {
    let index = this.repeatModes.indexOf(preferences.repeatMode) + 1

    if (index >= this.repeatModes.length) {
      index = 0
    }

    preferences.repeatMode = this.repeatModes[index]
  },

  /**
   * Play the prev song in the queue, if one is found.
   * If the prev song is not found and the current mode is NO_REPEAT, we stop completely.
   */
  playPrev (): void {
    // If the song's duration is greater than 5 seconds and we've passed 5 seconds into it,
    // restart playing instead.
    if (this.player!.media.currentTime > 5 && queueStore.current!.length > 5) {
      this.player!.restart()

      return
    }

    const prev = this.previous
    !prev && preferences.repeatMode === 'NO_REPEAT'
      ? this.stop()
      : this.play(prev)
  },

  /**
   * Play the next song in the queue, if one is found.
   * If the next song is not found and the current mode is NO_REPEAT, we stop completely.
   */
  playNext (): void {
    const next = this.next
    !next && preferences.repeatMode === 'NO_REPEAT'
      ? this.stop() //  Nothing lasts forever, even cold November rain.
      : this.play(next)
  },

  /**
   * @param {Number}     volume   0-10
   * @param {Boolean=true}   persist  Whether the volume should be saved into local storage
   */
  setVolume (volume: number, persist = true): void {
    this.player!.setVolume(volume)

    if (persist) {
      preferences.volume = volume
    }

    this.volumeInput!.value = String(volume)
  },

  mute (): void {
    this.setVolume(0, false)
  },

  unmute (): void {
    // If the saved volume is 0, we unmute to the default level (7).
    if (parseInt(preferences.volume) === 0) {
      preferences.volume = DEFAULT_VOLUME_VALUE
    }

    this.setVolume(preferences.volume)
  },

  stop () {
    document.title = app.name
    this.player!.pause()
    this.player!.seek(0)

    if (queueStore.current) {
      queueStore.current.playbackState = 'Stopped'
    }

    socket.broadcast(event.$names.SOCKET_PLAYBACK_STOPPED)
  },

  pause () {
    this.player!.pause()
    queueStore.current!.playbackState = 'Paused'
    socket.broadcast(event.$names.SOCKET_SONG, songStore.generateDataToBroadcast(queueStore.current!))
  },

  resume () {
    this.player!.play()
    queueStore.current!.playbackState = 'Playing'
    event.emit(event.$names.SONG_PLAYED, queueStore.current)
    socket.broadcast(event.$names.SOCKET_SONG, songStore.generateDataToBroadcast(queueStore.current!))
  },

  toggle () {
    if (!queueStore.current) {
      this.playFirstInQueue()
      return
    }

    if (queueStore.current.playbackState !== 'Playing') {
      this.resume()
      return
    }

    this.pause()
  },

  /**
   * Queue up songs (replace them into the queue) and start playing right away.
   *
   * @param {?Song[]} songs  An array of song objects. Defaults to all songs if null.
   * @param {Boolean=false}   shuffled Whether to shuffle the songs before playing.
   */
  queueAndPlay (songs?: Song[], shuffled: boolean = false): void {
    if (!songs) {
      songs = shuffle(songStore.all)
    }

    if (!songs.length) {
      return
    }

    if (shuffled) {
      songs = shuffle(songs)
    }

    queueStore.replaceQueueWith(songs)

    // Wrap this inside a nextTick() to wait for the DOM to complete updating
    // and then play the first song in the queue.
    Vue.nextTick((): void => {
      router.go('queue')
      this.play(queueStore.first)
    })
  },

  /**
   * Play the first song in the queue.
   * If the current queue is empty, try creating it by shuffling all songs.
   */
  playFirstInQueue (): void {
    queueStore.all.length ? this.play(queueStore.first) : this.queueAndPlay()
  },

  playAllByArtist ({ songs } : { songs: Song[] }, shuffled = true): void {
    shuffled
      ? this.queueAndPlay(songs, true /* shuffled */)
      : this.queueAndPlay(orderBy(songs, ['album_id', 'disc', 'track']))
  },

  playAllInAlbum ({ songs } : { songs: Song[]}, shuffled = true): void {
    shuffled
      ? this.queueAndPlay(songs, true /* shuffled */)
      : this.queueAndPlay(orderBy(songs, ['disc', 'track']))
  }
}
