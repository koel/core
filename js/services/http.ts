import Axios, { AxiosInstance, Method } from 'axios'
import NProgress from 'nprogress'

import { event } from '@/utils/index'
import { ls } from '@/services'

export const http = {
  client: null as AxiosInstance | null,

  request (method: Method, url: string, data: object, successCb: any, errorCb: any): void {
    this.client!.request({ url, data, method }).then(successCb).catch(errorCb)
  },

  get (url: string, successCb?: any, errorCb?: any): void {
    this.request('get', url, {}, successCb, errorCb)
  },

  post (url: string, data: object, successCb?: any, errorCb?: any): void {
    this.request('post', url, data, successCb, errorCb)
  },

  put (url: string, data: object, successCb?: any, errorCb?: any): void {
    this.request('put', url, data, successCb, errorCb)
  },

  delete (url: string, data: object = {}, successCb?: any, errorCb?: any): void {
    this.request('delete', url, data, successCb, errorCb)
  },

  init (): void {
    this.client = Axios.create({
      baseURL: KOEL_ENV === 'app' ? `${ls.get('koelHost')}api` : `${window.BASE_URL}api`
    })

    // Intercept the request to make sure the token is injected into the header.
    this.client.interceptors.request.use(config => {
      config.headers.Authorization = `Bearer ${ls.get('jwt-token')}`
      return config
    })

    // Intercept the response and…
    this.client.interceptors.response.use(response => {
      // …get the token from the header or response data if exists, and save it.
      const token = response.headers.Authorization || response.data.token

      if (token) {
        ls.set('jwt-token', token)
      }

      NProgress.done()

      return response
    }, error => {
      // Also, if we receive a Bad Request / Unauthorized error
      if (error.response.status === 400 || error.response.status === 401) {
        // and we're not trying to login
        if (!(error.config.method === 'post' && /\/api\/me\/?$/.test(error.config.url))) {
          // the token must have expired. Log out.
          event.emit(event.$names.LOG_OUT)
        }
      }

      NProgress.done()

      return Promise.reject(error)
    })
  }
}
