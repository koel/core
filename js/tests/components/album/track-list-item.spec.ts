import Component from '@/components/album/track-list-item.vue'
import { sharedStore, songStore, queueStore } from '@/stores'
import { playback, ls } from '@/services'
import factory from '@/tests/factory'
import { mock } from '@/tests/__helpers__'
import { shallow } from '@/tests/adapter'

describe('componnents/song/track-list-item', () => {
  let song: Song
  const track = {
    title: 'Foo and bar',
    fmtLenth: '00:42'
  }
  const album = factory('album', { id: 42 })
  window.BASE_URL = 'http://koel.local/'

  beforeEach(() => {
    sharedStore.state.useiTunes = true
    song = factory('song')
    mock(ls, 'get', 'abcdef')
  })

  afterEach(() => {
    jest.resetModules()
    jest.clearAllMocks()
  })

  it('renders', () => {
    const wrapper = shallow(Component, {
      propsData: {
        track,
        album,
        index: 1
      }
    })
    expect(wrapper).toMatchSnapshot()
  })

  it('plays', () => {
    mock(songStore, 'guess', song)
    const containsStub = mock(queueStore, 'contains', false)
    const queueStub = mock(queueStore, 'queueAfterCurrent')
    const playStub = mock(playback, 'play')

    shallow(Component, {
      propsData: {
        track,
        album,
        index: 1
      }
    }).click('li')

    expect(containsStub).toHaveBeenCalledWith(song)
    expect(queueStub).toHaveBeenCalledWith(song)
    expect(playStub).toHaveBeenCalledWith(song)
  })
})
