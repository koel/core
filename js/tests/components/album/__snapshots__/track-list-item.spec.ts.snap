// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`componnents/song/track-list-item renders 1`] = `
<li title="" role="button" tabindex="0" class=""><span class="no">2</span> <span class="title">Foo and bar</span> <a href="http://koel.local/api/itunes/song/42?q=Foo%20and%20bar&amp;jwt-token=abcdef" target="_blank" title="View on iTunes" class="view-on-itunes">
    iTunes
  </a> <span class="length"></span></li>
`;
